tests/csrf_tests/tests.py

```diff
@@ -694,3 +689,139 @@ def _set_post(self, post):
+    """
+    This is a subclass of CsrfViewMiddlewareTest that changes its behaviour to work with
+    CSRF_USE_SESSIONS = True to test the session-based CSRF handling in the same way as the
+    cookie-based handling.
+    """
+
+    def _get_POST_bare_secret_csrf_cookie_request(self):
+        req = self._get_POST_no_csrf_cookie_request()
+        req.session[settings.CSRF_COOKIE_NAME] = self._csrf_id_cookie[:32]
+        return req
+
+    def _get_GET_csrf_cookie_request(self):
+        req = TestingHttpRequest()
+        req.session[settings.CSRF_COOKIE_NAME] = self._csrf_id_cookie
+        return req
+
+    def setUp(self):
+        super(CsrfViewMiddlewareTest, self).setUp()
+
+        override = UserSettingsHolder(settings._wrapped)
+        override.CSRF_USE_SESSIONS = True
+        override.CSRF_COOKIE_DOMAIN = None
+        self.wrapped = settings._wrapped
+        settings._wrapped = override
+
+    def tearDown(self):
+        settings._wrapped = self.wrapped
+        del self.wrapped
+        super(CsrfViewMiddlewareTest, self).tearDown()
+
+    def test_process_response_get_token_used(self):
+        """
+        The ensure_csrf_cookie() decorator works without middleware.
+        """
+        @ensure_csrf_cookie
+        def view(request):
+            # Doesn't insert a token or anything
+            return HttpResponse(content="")
+
+        req = self._get_GET_no_csrf_cookie_request()
+        view(req)
+        self.assertTrue(req.session.get(settings.CSRF_SESSION_KEY, False))
+
+    def test_ensures_csrf_cookie_with_middleware(self):
+        """
+        The ensure_csrf_cookie() decorator works with the CsrfViewMiddleware
+        enabled.
+        """
+        @ensure_csrf_cookie
+        def view(request):
+            # Doesn't insert a token or anything
+            return HttpResponse(content="")
+
+        req = self._get_GET_no_csrf_cookie_request()
+        CsrfViewMiddleware().process_view(req, view, (), {})
+        resp = view(req)
+        CsrfViewMiddleware().process_response(req, resp)
+        self.assertTrue(req.session.get(settings.CSRF_SESSION_KEY, False))
+
+    def test_ensures_csrf_cookie_no_middleware(self):
+        pass  # Does not apply to sessions
+
+    def test_process_view_token_invalid_bytes(self):
+        pass  # Does not apply to sessions
+
+    def test_process_view_token_invalid_chars(self):
+        pass  # Does not apply to sessions
+
+    def test_process_view_token_too_long(self):
+        pass  # Does not apply to sessions
+
+    def test_bare_secret_accepted_and_replaced(self):
+        # Does not apply to sessions, as CSRF tokens in sessions are added in 1.11 and bare secrets
+        # vanished in 1.10, so we cannot have bare secrets in a session
+        pass
+
+    def test_csrf_cookie_age(self):
+        pass  # Does not apply to sessions
+
+    def test_csrf_cookie_age_none(self):
+        pass  # Does not apply to sessions
+
```

Comment:
All of these empty tests look awkward. I'll try to find a better way to share some tests -- probably along the lines of
```python
class CSRFTestBase(object):
    # all the common parts go here

class CSRFCookieTest(CSRFTestBase, TestCase):
    # Cookie-specific tests and utilities

class CSRFSessionTest(CSRFTestBase, TestCase):
    # Session-specific tests and utilities